index.ts

```diff
@@ -0,0 +1,245 @@
+const EthQuery = require('ethjs-query')
+import assert from 'assert';
+const Mutex = require('await-semaphore').Mutex
+
+/**
+    @property opts.provider - A ethereum provider
+    @property opts.blockTracker - An instance of eth-blocktracker
+    @property opts.getPendingTransactions - A function that returns an array of txMeta
+    whosee status is `submitted`
+    @property opts.getConfirmedTransactions - A function that returns an array of txMeta
+    whose status is `confirmed`
+*/
+export interface opts {
+  provider: object;
+  blockTracker: any;
+  getPendingTransactions(address: string):Transactions[];
+  getConfirmedTransactions(address: string): Transactions[];
+}
+
+/**
+  * @property  highestLocallyConfirmed - A hex string of the highest nonce on a confirmed transaction.
+  * @property  nextNetworkNonce - The next nonce suggested by the eth_getTransactionCount method.
+  * @property  highestSuggested - The maximum between the other two, the number returned.
+  * @property  local - Nonce details derived from pending transactions and highestSuggested
+  * @property  network - Nonce details from the eth_getTransactionCount method
+*/
+export interface NonceDetails {
+  params:{
+  highestLocallyConfirmed: number;
+  nextNetworkNonce: number;
+  highestSuggested: number;
+  };
+  local: HighestContinuousFrom;
+  network: NetWorkNextNonce;
+}
+
+/**
+  * @property  nextNonce - The highest of the nonce values derived based on confirmed and pending transactions and eth_getTransactionCount method
+  * @property  nonceDetails - details of nonce value derivation.
+  * @property  releaseLock 
+*/
+export interface NonceLock{ 
+  nextNonce: number; 
+  nonceDetails: NonceDetails; 
+  releaseLock: any;
+}
+
+/**
+  * @property  name - The name for how the nonce was calculated based on the data used
+  * @property  nonce - The next nonce value suggested by the eth_getTransactionCount method.
+  * @property  blockNumber - The latest block from the network
+  * @property  baseCount - Transaction count from the network suggested by eth_getTransactionCount method
+*/
+export interface NetWorkNextNonce{ 
+  name: string; 
+  nonce: number; 
+  details: { 
+    blockNumber: Promise<any>; 
+    baseCount: number;
+  }
+}
+
+/**
+  @property name- The name for how the nonce was calculated based on the data used
+  @property nonce - the next suggested nonce
+  @property details{startPoint, highest} - the provided starting nonce that was used and highest derived from it (for debugging)
+*/
+export interface HighestContinuousFrom{
+  name: string; 
+  nonce: number; 
+  details: {
+    startPoint: number;
+    highest: number;
+  }
+}
+export interface Transactions {
+  status: string;
+  history: [{}];
```

Comment:
For plain objects, we should use the following:
```suggestion
  history: [Record<string, unknown>];
```

Reply:
I was a little confused about how to initialize plain objects. I was having trouble with initializing `lockMap` variable as well. Thank you.